File: GhostBot/functions/buffs.py
from __future__ import annotations

import time

from typing import TYPE_CHECKING

from GhostBot.config import BuffConfig
from GhostBot.functions.runner import Runner, run_at_interval
from GhostBot.lib.math import seconds

if TYPE_CHECKING:
    from GhostBot.bot_controller import ExtendedClient


@run_at_interval()
class Buffs(Runner):

    def __init__(self, client: ExtendedClient):
        super().__init__(client)
        self.config: BuffConfig = client.config.buff
        self._interval = seconds(minutes=int(self.config.interval))

    def _run(self) -> None:
        self._log_info(f'Buffing.')
        for _key, _sleep in self.config.buffs:
            self._client.press_key(_key)
            time.sleep(seconds(seconds=_sleep))

File: kotlin-server/src/main/kotlin/com/ghostbot/functions/Buffs.kt
package com.ghostbot.functions

import com.ghostbot.controller.BotClientWindow
import com.ghostbot.lib.seconds
import com.ghostbot.lib.sleepSeconds

/**
 * Port of GhostBot/functions/buffs.py — `@run_at_interval(run_on_start=True)`.
 */
class Buffs(client: BotClientWindow) : Runner(client) {

    private val config = client.config?.buff ?: error("buff config required")

    override val runOnStart: Boolean get() = true
    override val intervalMs: Long = seconds(minutes = (config.interval as? Int) ?: 10).toLong() * 1000

    override fun _run(): Boolean {
        _logInfo("Buffing.")
        for (row in config.buffs ?: emptyList()) {
            if (row.size < 2) continue
            client.pressKey(row[0])
            sleepSeconds(((row[1] as? Number)?.toDouble() ?: 0.0) / 1000.0)
        }
        return true
    }
}
